py/binary.c

```diff
@@ -303,7 +303,12 @@ void mp_binary_set_val(char struct_type, char val_type, mp_obj_t val_in, byte **
                 // zero/sign extend if needed
                 if (BYTES_PER_WORD < 8 && size > sizeof(val)) {
                     int c = (is_signed(val_type) && (mp_int_t)val < 0) ? 0xff : 0x00;
-                    memset(p + sizeof(val), c, size - sizeof(val));
+                    if (struct_type == '>') {
+                        memset(p, c, size - sizeof(val));
+                        p += size - sizeof(val);
+                    } else {
+                        memset(p + sizeof(val), c, size - sizeof(val));
+                    }
```

Comment:
There is a chance to optimise this to reduce code size, by having just a single memset that fills the entire buffer pointed to by p, ie memset(p, c, size).  Since size is limited to a maximum of 8 bytes this memset is small and bounded so there's very little performance that would be lost due to filling the whole buffer.  And the case of using the 'Q' typecode is relatively rare so I'd say it's fair to optimise this case for code size, not speed.